### PR TITLE
sync 1.1.4 >> feature 1.2 [full merge, no squash]

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Update release branch with latest commits from master.

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
pls, do not merge the target branch, this shall be always rebase only...

**Comment by SeanNaren** [comment]
> pls, do not merge the target branch, this shall be always rebase only...

Should I drop the commits/merge?

**Comment by Borda** [comment]
> > pls, do not merge the target branch, this shall be always rebase only...
> 
> Should I drop the commits/merge?

I would say yes, otherwise, it can bite us later when merging mack to master
(if we would do squash-merge as all other PRs then this merge does not matter but with full merge, it can be a problem)

**Comment by SeanNaren** [comment]
Done @Borda, just did a rebase on top of the 1.2-dev branch.

File: CHANGELOG.md [hunk 1]
```diff
@@ -94,7 +94,29 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+
+### Deprecated
+
+
+### Removed
+
+
+### Fixed
+
+- Fixed `transfer_batch_to_device` for DDP with `len(devices_ids) == 1` ([#5195](https://github.com/PyTorchLightning/pytorch-lightning/pull/5195))
```
Comment (Borda) [inline CHANGELOG.md:115]: this looks very sparse...

File: pytorch_lightning/utilities/package_utils.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        mods = module_path.split('.')
+        assert mods, 'nothing given to test'
+        # it has to be tested as per partets
+        for i in range(len(mods)):
+            module_path = '.'.join(mods[:i + 1])
+            if importlib.util.find_spec(module_path) is None:
+                return False
+        return True
+    except AttributeError:
+        return False
```
Comment (Borda) [inline pytorch_lightning/utilities/package_utils.py:36]: this is already in pytorch_lightning/utilities/imports.py

File: pytorch_lightning/utilities/parsing.py [hunk 3]
```diff
@@ -15,9 +15,11 @@
 import inspect
 import pickle
 from argparse import Namespace
-from typing import Dict, Union, Tuple
+from typing import Dict, Tuple, Union
 
 from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.package_utils import _module_available
```
Comment (Borda) [inline pytorch_lightning/utilities/parsing.py:22]:
```suggestion
from pytorch_lightning.utilities.imports import _module_available
```

**Comment by SeanNaren** [comment]
@carmocca is setting up a new PR with 1.1.5 changes as well.

**Review by Borda — APPROVED** [review]

**Comment by SeanNaren** [comment]
@Borda why did you re-open this? https://github.com/PyTorchLightning/pytorch-lightning/pull/5583 contains the commits from this sync

**Comment by Borda** [comment]
since #5583 seems to be failing, let's merge this asap and then fix the update with 1.1.5
cc: @tchaton @williamFalcon @justusschock @awaelchli @edenlightning

**Comment by Borda** [comment]
fyi, the failing `mypy` is kind of fixed in #5634

**Comment by Borda** [comment]
Removing ready to go label just to avoid bot's updating this PR...

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Great work !

**Review by edenlightning — APPROVED** [review]

**Pull request merged**